// secretjs/dist/tx/authz.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgRevoke = exports.MsgExec = exports.MsgGrant = exports.StakeAuthorizationType = exports.MsgGrantAuthorization = void 0;
var MsgGrantAuthorization;
(function (MsgGrantAuthorization) {
    MsgGrantAuthorization["MsgAcknowledgement"] = "/ibc.core.channel.v1.MsgAcknowledgement";
    MsgGrantAuthorization["MsgBeginRedelegate"] = "/cosmos.staking.v1beta1.MsgBeginRedelegate";
    MsgGrantAuthorization["MsgChannelCloseConfirm"] = "/ibc.core.channel.v1.MsgChannelCloseConfirm";
    MsgGrantAuthorization["MsgChannelCloseInit"] = "/ibc.core.channel.v1.MsgChannelCloseInit";
    MsgGrantAuthorization["MsgChannelOpenAck"] = "/ibc.core.channel.v1.MsgChannelOpenAck";
    MsgGrantAuthorization["MsgChannelOpenConfirm"] = "/ibc.core.channel.v1.MsgChannelOpenConfirm";
    MsgGrantAuthorization["MsgChannelOpenInit"] = "/ibc.core.channel.v1.MsgChannelOpenInit";
    MsgGrantAuthorization["MsgChannelOpenTry"] = "/ibc.core.channel.v1.MsgChannelOpenTry";
    MsgGrantAuthorization["MsgConnectionOpenAck"] = "/ibc.core.connection.v1.MsgConnectionOpenAck";
    MsgGrantAuthorization["MsgConnectionOpenConfirm"] = "/ibc.core.connection.v1.MsgConnectionOpenConfirm";
    MsgGrantAuthorization["MsgConnectionOpenInit"] = "/ibc.core.connection.v1.MsgConnectionOpenInit";
    MsgGrantAuthorization["MsgConnectionOpenTry"] = "/ibc.core.connection.v1.MsgConnectionOpenTry";
    MsgGrantAuthorization["MsgCreateClient"] = "/ibc.core.client.v1.MsgCreateClient";
    MsgGrantAuthorization["MsgCreateValidator"] = "/cosmos.staking.v1beta1.MsgCreateValidator";
    MsgGrantAuthorization["MsgDelegate"] = "/cosmos.staking.v1beta1.MsgDelegate";
    MsgGrantAuthorization["MsgDeposit"] = "/cosmos.gov.v1beta1.MsgDeposit";
    MsgGrantAuthorization["MsgEditValidator"] = "/cosmos.staking.v1beta1.MsgEditValidator";
    MsgGrantAuthorization["MsgExec"] = "/cosmos.authz.v1beta1.MsgExec";
    MsgGrantAuthorization["MsgExecuteContract"] = "/secret.compute.v1beta1.MsgExecuteContract";
    MsgGrantAuthorization["MsgFundCommunityPool"] = "/cosmos.distribution.v1beta1.MsgFundCommunityPool";
    MsgGrantAuthorization["MsgGrant"] = "/cosmos.authz.v1beta1.MsgGrant";
    MsgGrantAuthorization["MsgGrantAllowance"] = "/cosmos.feegrant.v1beta1.MsgGrantAllowance";
    MsgGrantAuthorization["MsgInstantiateContract"] = "/secret.compute.v1beta1.MsgInstantiateContract";
    MsgGrantAuthorization["MsgMultiSend"] = "/cosmos.bank.v1beta1.MsgMultiSend";
    MsgGrantAuthorization["MsgRecvPacket"] = "/ibc.core.channel.v1.MsgRecvPacket";
    MsgGrantAuthorization["MsgRevoke"] = "/cosmos.authz.v1beta1.MsgRevoke";
    MsgGrantAuthorization["MsgRevokeAllowance"] = "/cosmos.feegrant.v1beta1.MsgRevokeAllowance";
    MsgGrantAuthorization["MsgSend"] = "/cosmos.bank.v1beta1.MsgSend";
    MsgGrantAuthorization["MsgSetWithdrawAddress"] = "/cosmos.distribution.v1beta1.MsgSetWithdrawAddress";
    MsgGrantAuthorization["MsgStoreCode"] = "/secret.compute.v1beta1.MsgStoreCode";
    MsgGrantAuthorization["MsgSubmitEvidence"] = "/cosmos.evidence.v1beta1.MsgSubmitEvidence";
    MsgGrantAuthorization["MsgSubmitMisbehaviour"] = "/ibc.core.client.v1.MsgSubmitMisbehaviour";
    MsgGrantAuthorization["MsgSubmitProposal"] = "/cosmos.gov.v1beta1.MsgSubmitProposal";
    MsgGrantAuthorization["MsgTimeout"] = "/ibc.core.channel.v1.MsgTimeout";
    MsgGrantAuthorization["MsgTimeoutOnClose"] = "/ibc.core.channel.v1.MsgTimeoutOnClose";
    MsgGrantAuthorization["MsgTransfer"] = "/ibc.applications.transfer.v1.MsgTransfer";
    MsgGrantAuthorization["MsgUndelegate"] = "/cosmos.staking.v1beta1.MsgUndelegate";
    MsgGrantAuthorization["MsgUnjail"] = "/cosmos.slashing.v1beta1.MsgUnjail";
    MsgGrantAuthorization["MsgUpdateClient"] = "/ibc.core.client.v1.MsgUpdateClient";
    MsgGrantAuthorization["MsgUpgradeClient"] = "/ibc.core.client.v1.MsgUpgradeClient";
    MsgGrantAuthorization["MsgVerifyInvariant"] = "/cosmos.crisis.v1beta1.MsgVerifyInvariant";
    MsgGrantAuthorization["MsgVote"] = "/cosmos.gov.v1beta1.MsgVote";
    MsgGrantAuthorization["MsgVoteWeighted"] = "/cosmos.gov.v1beta1.MsgVoteWeighted";
    MsgGrantAuthorization["MsgWithdrawDelegatorReward"] = "/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward";
    MsgGrantAuthorization["MsgWithdrawValidatorCommission"] = "/cosmos.distribution.v1beta1.MsgWithdrawValidatorCommission";
})(MsgGrantAuthorization = exports.MsgGrantAuthorization || (exports.MsgGrantAuthorization = {}));
function isGenericAuthorization(object) {
    return "msg" in object;
}
function isSendAuthorization(object) {
    return "spendLimit" in object;
}
function isStakeAuthorization(object) {
    return ("maxTokens" in object &&
        "allowList" in object &&
        "denyList" in object &&
        "authorizationType" in object);
}
/** AuthorizationType defines the type of staking module authorization type */
var StakeAuthorizationType;
(function (StakeAuthorizationType) {
    /** defines an authorization type for Msg/Delegate */
    StakeAuthorizationType[StakeAuthorizationType["Delegate"] = 1] = "Delegate";
    /** defines an authorization type for Msg/Undelegate */
    StakeAuthorizationType[StakeAuthorizationType["Undelegate"] = 2] = "Undelegate";
    /** defines an authorization type for Msg/BeginRedelegate */
    StakeAuthorizationType[StakeAuthorizationType["Redelegate"] = 3] = "Redelegate";
})(StakeAuthorizationType = exports.StakeAuthorizationType || (exports.StakeAuthorizationType = {}));
/**
 * MsgGrant is a request type for Grant method. It declares authorization to the grantee
 * on behalf of the granter with the provided expiration time.
 */
class MsgGrant {
    constructor(params) {
        this.params = params;
    }
    toProto() {
        return __awaiter(this, void 0, void 0, function* () {
            let grant;
            if (isSendAuthorization(this.params.authorization)) {
                const { SendAuthorization } = yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/bank/v1beta1/authz")));
                grant = {
                    authorization: {
                        typeUrl: "/cosmos.bank.v1beta1.SendAuthorization",
                        value: SendAuthorization.encode(this.params.authorization).finish(),
                    },
                    expiration: { seconds: String(this.params.expiration), nanos: 0 },
                };
            }
            else if (isStakeAuthorization(this.params.authorization)) {
                const { StakeAuthorization } = yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/staking/v1beta1/authz")));
                grant = {
                    authorization: {
                        typeUrl: "/cosmos.staking.v1beta1.StakeAuthorization",
                        value: StakeAuthorization.encode({
                            maxTokens: this.params.authorization.maxTokens,
                            allowList: { address: this.params.authorization.allowList },
                            denyList: { address: this.params.authorization.denyList },
                            authorizationType: Number(this.params.authorization.authorizationType),
                        }).finish(),
                    },
                    expiration: { seconds: String(this.params.expiration), nanos: 0 },
                };
            }
            else if (isGenericAuthorization(this.params.authorization)) {
                const { GenericAuthorization } = yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/authz/v1beta1/authz")));
                grant = {
                    authorization: {
                        typeUrl: "/cosmos.authz.v1beta1.GenericAuthorization",
                        value: GenericAuthorization.encode({
                            msg: String(this.params.authorization.msg),
                        }).finish(),
                    },
                    expiration: { seconds: String(this.params.expiration), nanos: 0 },
                };
            }
            else {
                throw new Error("Unknown authorization type.");
            }
            const msgContent = {
                granter: this.params.granter,
                grantee: this.params.grantee,
                grant: grant,
            };
            return {
                typeUrl: "/cosmos.authz.v1beta1.MsgGrant",
                value: msgContent,
                encode: () => __awaiter(this, void 0, void 0, function* () {
                    return (yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/authz/v1beta1/tx")))).MsgGrant.encode(msgContent).finish();
                }),
            };
        });
    }
    toAmino() {
        return __awaiter(this, void 0, void 0, function* () {
            throw new Error("MsgGrant not implemented.");
        });
    }
}
exports.MsgGrant = MsgGrant;
/**
 * MsgExec attempts to execute the provided messages using
 * authorizations granted to the grantee. Each message should have only
 * one signer corresponding to the granter of the authorization.
 */
class MsgExec {
    constructor(params) {
        this.params = params;
    }
    toProto(utils) {
        return __awaiter(this, void 0, void 0, function* () {
            const msgContent = {
                grantee: this.params.grantee,
                msgs: yield Promise.all(this.params.msgs.map((m) => m.toProto(utils))),
            };
            return {
                typeUrl: "/cosmos.authz.v1beta1.MsgExec",
                value: msgContent,
                encode: () => __awaiter(this, void 0, void 0, function* () {
                    return (yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/authz/v1beta1/tx")))).MsgExec.encode(msgContent).finish();
                }),
            };
        });
    }
    toAmino() {
        return __awaiter(this, void 0, void 0, function* () {
            throw new Error("MsgExec not implemented.");
        });
    }
}
exports.MsgExec = MsgExec;
/**
 * MsgRevoke revokes any authorization with the provided sdk.Msg type on the
 * granter's account with that has been granted to the grantee.
 */
class MsgRevoke {
    constructor(params) {
        this.params = params;
    }
    toProto() {
        return __awaiter(this, void 0, void 0, function* () {
            const msgContent = {
                granter: this.params.granter,
                grantee: this.params.grantee,
                msgTypeUrl: String(this.params.msg),
            };
            return {
                typeUrl: "/cosmos.authz.v1beta1.MsgRevoke",
                value: msgContent,
                encode: () => __awaiter(this, void 0, void 0, function* () {
                    return (yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/authz/v1beta1/tx")))).MsgRevoke.encode(msgContent).finish();
                }),
            };
        });
    }
    toAmino() {
        return __awaiter(this, void 0, void 0, function* () {
            throw new Error("MsgRevoke not implemented.");
        });
    }
}
exports.MsgRevoke = MsgRevoke;
