// secretjs/dist/tx/compute.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgStoreCode = exports.MsgExecuteContract = exports.MsgInstantiateContract = exports.getMissingCodeHashWarning = void 0;
const encoding_1 = require("@cosmjs/encoding");
const is_gzip_1 = __importDefault(require("is-gzip"));
const compute_1 = require("../query/compute");
function getMissingCodeHashWarning(method) {
    return `${new Date().toISOString()} WARNING: ${method} was used without the "codeHash" parameter. This is discouraged and will result in much slower execution times for your app.`;
}
exports.getMissingCodeHashWarning = getMissingCodeHashWarning;
/** Instantiate a contract from code id */
class MsgInstantiateContract {
    constructor({ sender, codeId, label, initMsg, initFunds, codeHash, }) {
        this.warnCodeHash = false;
        this.sender = sender;
        this.codeId = String(codeId);
        this.label = label;
        this.initMsg = initMsg;
        this.initMsgEncrypted = null;
        this.initFunds = initFunds !== null && initFunds !== void 0 ? initFunds : [];
        if (codeHash) {
            this.codeHash = codeHash.replace("0x", "").toLowerCase();
        }
        else {
            // codeHash will be set in SecretNetworkClient before invoking toProto() & toAimno()
            this.codeHash = "";
            this.warnCodeHash = true;
            console.warn(getMissingCodeHashWarning("MsgInstantiateContract"));
        }
    }
    toProto(utils) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.warnCodeHash) {
                console.warn(getMissingCodeHashWarning("MsgInstantiateContract"));
            }
            if (!this.initMsgEncrypted) {
                // The encryption uses a random nonce
                // toProto() & toAmino() are called multiple times during signing
                // so to keep the msg consistant across calls we encrypt the msg only once
                this.initMsgEncrypted = yield utils.encrypt(this.codeHash, this.initMsg);
            }
            const msgContent = {
                sender: (0, compute_1.addressToBytes)(this.sender),
                codeId: this.codeId,
                label: this.label,
                initMsg: this.initMsgEncrypted,
                initFunds: this.initFunds,
                // callbackSig & callbackCodeHash are internal stuff that doesn't matter here
                callbackSig: new Uint8Array(),
                callbackCodeHash: "",
            };
            return {
                typeUrl: "/secret.compute.v1beta1.MsgInstantiateContract",
                value: msgContent,
                encode: () => __awaiter(this, void 0, void 0, function* () {
                    return (yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/secret/compute/v1beta1/msg")))).MsgInstantiateContract.encode(msgContent).finish();
                }),
            };
        });
    }
    toAmino(utils) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.warnCodeHash) {
                console.warn(getMissingCodeHashWarning("MsgInstantiateContract"));
            }
            if (!this.initMsgEncrypted) {
                // The encryption uses a random nonce
                // toProto() & toAmino() are called multiple times during signing
                // so to keep the msg consistant across calls we encrypt the msg only once
                this.initMsgEncrypted = yield utils.encrypt(this.codeHash, this.initMsg);
            }
            return {
                type: "wasm/MsgInstantiateContract",
                value: {
                    sender: this.sender,
                    code_id: this.codeId,
                    label: this.label,
                    init_msg: (0, encoding_1.toBase64)(this.initMsgEncrypted),
                    init_funds: this.initFunds,
                },
            };
        });
    }
}
exports.MsgInstantiateContract = MsgInstantiateContract;
/** Execute a function on a contract */
class MsgExecuteContract {
    constructor({ sender, contractAddress, msg, sentFunds, codeHash, }) {
        this.warnCodeHash = false;
        this.sender = sender;
        this.contractAddress = contractAddress;
        this.msg = msg;
        this.msgEncrypted = null;
        this.sentFunds = sentFunds !== null && sentFunds !== void 0 ? sentFunds : [];
        if (codeHash) {
            this.codeHash = codeHash.replace("0x", "").toLowerCase();
        }
        else {
            // codeHash will be set in SecretNetworkClient before invoking toProto() & toAimno()
            this.codeHash = "";
            this.warnCodeHash = true;
            console.warn(getMissingCodeHashWarning("MsgExecuteContract"));
        }
    }
    toProto(utils) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.warnCodeHash) {
                console.warn(getMissingCodeHashWarning("MsgExecuteContract"));
            }
            if (!this.msgEncrypted) {
                // The encryption uses a random nonce
                // toProto() & toAmino() are called multiple times during signing
                // so to keep the msg consistant across calls we encrypt the msg only once
                this.msgEncrypted = yield utils.encrypt(this.codeHash, this.msg);
            }
            const msgContent = {
                sender: (0, compute_1.addressToBytes)(this.sender),
                contract: (0, compute_1.addressToBytes)(this.contractAddress),
                msg: this.msgEncrypted,
                sentFunds: this.sentFunds,
                // callbackSig & callbackCodeHash are internal stuff that doesn't matter here
                callbackSig: new Uint8Array(),
                callbackCodeHash: "",
            };
            return {
                typeUrl: "/secret.compute.v1beta1.MsgExecuteContract",
                value: msgContent,
                encode: () => __awaiter(this, void 0, void 0, function* () {
                    return (yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/secret/compute/v1beta1/msg")))).MsgExecuteContract.encode(msgContent).finish();
                }),
            };
        });
    }
    toAmino(utils) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.warnCodeHash) {
                console.warn(getMissingCodeHashWarning("MsgExecuteContract"));
            }
            if (!this.msgEncrypted) {
                // The encryption uses a random nonce
                // toProto() & toAmino() are called multiple times during signing
                // so to keep the msg consistant across calls we encrypt the msg only once
                this.msgEncrypted = yield utils.encrypt(this.codeHash, this.msg);
            }
            return {
                type: "wasm/MsgExecuteContract",
                value: {
                    sender: this.sender,
                    contract: this.contractAddress,
                    msg: (0, encoding_1.toBase64)(this.msgEncrypted),
                    sent_funds: this.sentFunds,
                },
            };
        });
    }
}
exports.MsgExecuteContract = MsgExecuteContract;
/** Upload a compiled contract to Secret Network */
class MsgStoreCode {
    constructor({ sender, wasmByteCode, source, builder }) {
        this.sender = sender;
        this.source = source;
        this.builder = builder;
        this.wasmByteCode = wasmByteCode;
    }
    gzipWasm() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!(0, is_gzip_1.default)(this.wasmByteCode)) {
                const pako = yield Promise.resolve().then(() => __importStar(require("pako")));
                this.wasmByteCode = pako.gzip(this.wasmByteCode, { level: 9 });
            }
        });
    }
    toProto() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.gzipWasm();
            const msgContent = {
                sender: (0, compute_1.addressToBytes)(this.sender),
                wasmByteCode: this.wasmByteCode,
                source: this.source,
                builder: this.builder,
            };
            return {
                typeUrl: "/secret.compute.v1beta1.MsgStoreCode",
                value: msgContent,
                encode: () => __awaiter(this, void 0, void 0, function* () {
                    return (yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/secret/compute/v1beta1/msg")))).MsgStoreCode.encode(msgContent).finish();
                }),
            };
        });
    }
    toAmino() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.gzipWasm();
            return {
                type: "wasm/MsgStoreCode",
                value: {
                    sender: this.sender,
                    wasm_byte_code: (0, encoding_1.toBase64)(this.wasmByteCode),
                    source: this.source.length > 0 ? this.source : undefined,
                    builder: this.builder.length > 0 ? this.builder : undefined,
                },
            };
        });
    }
}
exports.MsgStoreCode = MsgStoreCode;
