// secretjs/dist/tx/staking.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUndelegate = exports.MsgBeginRedelegate = exports.MsgDelegate = exports.MsgEditValidator = exports.MsgCreateValidator = void 0;
const encoding_1 = require("@cosmjs/encoding");
const bech32_1 = require("bech32");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
/** MsgCreateValidator defines an SDK message for creating a new validator. */
class MsgCreateValidator {
    constructor({ description, commission, minSelfDelegation, selfDelegatorAddress, pubkey, initialDelegation, }) {
        this.description = description;
        this.commission = commission;
        this.minSelfDelegation = minSelfDelegation;
        this.delegatorAddress = selfDelegatorAddress;
        this.validatorAddress = bech32_1.bech32.encode("secretvaloper", bech32_1.bech32.decode(selfDelegatorAddress).words);
        this.pubkey = pubkey;
        this.initialDelegation = initialDelegation;
    }
    toProto() {
        return __awaiter(this, void 0, void 0, function* () {
            const { PubKey } = yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/crypto/ed25519/keys")));
            const { Any } = yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/google/protobuf/any")));
            const msgContent = {
                description: this.description,
                commission: {
                    rate: new bignumber_js_1.default(this.commission.rate)
                        .toFixed(18)
                        .replace(/0\.0*/, ""),
                    maxRate: new bignumber_js_1.default(this.commission.maxRate)
                        .toFixed(18)
                        .replace(/0\.0*/, ""),
                    maxChangeRate: new bignumber_js_1.default(this.commission.maxChangeRate)
                        .toFixed(18)
                        .replace(/0\.0*/, ""),
                },
                minSelfDelegation: this.minSelfDelegation,
                delegatorAddress: this.delegatorAddress,
                validatorAddress: this.validatorAddress,
                pubkey: Any.fromPartial({
                    typeUrl: "/cosmos.crypto.ed25519.PubKey",
                    value: PubKey.encode(PubKey.fromPartial({
                        key: (0, encoding_1.fromBase64)(this.pubkey),
                    })).finish(),
                }),
                value: this.initialDelegation,
            };
            return {
                typeUrl: `/cosmos.staking.v1beta1.MsgCreateValidator`,
                value: msgContent,
                encode: () => __awaiter(this, void 0, void 0, function* () {
                    return (yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/staking/v1beta1/tx")))).MsgCreateValidator.encode(msgContent).finish();
                }),
            };
        });
    }
    toAmino() {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                type: "cosmos-sdk/MsgCreateValidator",
                value: {
                    description: {
                        moniker: this.description.moniker,
                        identity: this.description.identity,
                        website: this.description.website,
                        security_contact: this.description.securityContact,
                        details: this.description.details,
                    },
                    commission: {
                        rate: new bignumber_js_1.default(this.commission.rate).toFixed(18),
                        max_rate: new bignumber_js_1.default(this.commission.maxRate).toFixed(18),
                        max_change_rate: new bignumber_js_1.default(this.commission.maxChangeRate).toFixed(18),
                    },
                    min_self_delegation: this.minSelfDelegation,
                    delegator_address: this.delegatorAddress,
                    validator_address: this.validatorAddress,
                    pubkey: {
                        type: "tendermint/PubKeyEd25519",
                        value: this.pubkey,
                    },
                    value: this.initialDelegation,
                },
            };
        });
    }
}
exports.MsgCreateValidator = MsgCreateValidator;
/** MsgEditValidator defines an SDK message for editing an existing validator. */
class MsgEditValidator {
    constructor({ description, validatorAddress, commissionRate, minSelfDelegation, }) {
        this.validatorAddress = validatorAddress;
        this.description = description;
        this.commissionRate = commissionRate;
        this.minSelfDelegation = minSelfDelegation;
    }
    toProto() {
        return __awaiter(this, void 0, void 0, function* () {
            const { Description } = yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/staking/v1beta1/staking")));
            const msgContent = {
                validatorAddress: this.validatorAddress,
                description: Description.fromPartial(this.description || {}),
                commissionRate: this.commissionRate
                    ? new bignumber_js_1.default(this.commissionRate).toFixed(18).replace(/0\.0*/, "")
                    : "",
                minSelfDelegation: this.minSelfDelegation || "",
            };
            return {
                typeUrl: `/cosmos.staking.v1beta1.MsgEditValidator`,
                value: msgContent,
                encode: () => __awaiter(this, void 0, void 0, function* () {
                    return (yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/staking/v1beta1/tx")))).MsgEditValidator.encode(msgContent).finish();
                }),
            };
        });
    }
    toAmino() {
        return __awaiter(this, void 0, void 0, function* () {
            let description;
            if (this.description) {
                description = {
                    moniker: this.description.moniker,
                    identity: this.description.identity,
                    website: this.description.website,
                    security_contact: this.description.securityContact,
                    details: this.description.details,
                };
            }
            let commission_rate;
            if (this.commissionRate) {
                commission_rate = new bignumber_js_1.default(this.commissionRate).toFixed(18);
            }
            return {
                type: "cosmos-sdk/MsgEditValidator",
                value: {
                    validator_address: this.validatorAddress,
                    description,
                    commission_rate,
                    min_self_delegation: this.minSelfDelegation,
                },
            };
        });
    }
}
exports.MsgEditValidator = MsgEditValidator;
/** MsgDelegate defines an SDK message for performing a delegation of coins from a delegator to a validator. */
class MsgDelegate {
    constructor({ delegatorAddress, validatorAddress, amount, }) {
        this.delegatorAddress = delegatorAddress;
        this.validatorAddress = validatorAddress;
        this.amount = amount;
    }
    toProto() {
        return __awaiter(this, void 0, void 0, function* () {
            const msgContent = {
                delegatorAddress: this.delegatorAddress,
                validatorAddress: this.validatorAddress,
                amount: this.amount,
            };
            return {
                typeUrl: `/cosmos.staking.v1beta1.MsgDelegate`,
                value: msgContent,
                encode: () => __awaiter(this, void 0, void 0, function* () {
                    return (yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/staking/v1beta1/tx")))).MsgDelegate.encode(msgContent).finish();
                }),
            };
        });
    }
    toAmino() {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                type: "cosmos-sdk/MsgDelegate",
                value: {
                    delegator_address: this.delegatorAddress,
                    validator_address: this.validatorAddress,
                    amount: this.amount,
                },
            };
        });
    }
}
exports.MsgDelegate = MsgDelegate;
/** MsgBeginRedelegate defines an SDK message for performing a redelegation of coins from a delegator and source validator to a destination validator. */
class MsgBeginRedelegate {
    constructor({ delegatorAddress, validatorSrcAddress, validatorDstAddress, amount, }) {
        this.delegatorAddress = delegatorAddress;
        this.validatorSrcAddress = validatorSrcAddress;
        this.validatorDstAddress = validatorDstAddress;
        this.amount = amount;
    }
    toProto() {
        return __awaiter(this, void 0, void 0, function* () {
            const msgContent = {
                delegatorAddress: this.delegatorAddress,
                validatorSrcAddress: this.validatorSrcAddress,
                validatorDstAddress: this.validatorDstAddress,
                amount: this.amount,
            };
            return {
                typeUrl: `/cosmos.staking.v1beta1.MsgBeginRedelegate`,
                value: msgContent,
                encode: () => __awaiter(this, void 0, void 0, function* () {
                    return (yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/staking/v1beta1/tx")))).MsgBeginRedelegate.encode(msgContent).finish();
                }),
            };
        });
    }
    toAmino() {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                type: "cosmos-sdk/MsgBeginRedelegate",
                value: {
                    delegator_address: this.delegatorAddress,
                    validator_src_address: this.validatorSrcAddress,
                    validator_dst_address: this.validatorDstAddress,
                    amount: this.amount,
                },
            };
        });
    }
}
exports.MsgBeginRedelegate = MsgBeginRedelegate;
/** MsgUndelegate defines an SDK message for performing an undelegation from a delegate and a validator */
class MsgUndelegate {
    constructor({ delegatorAddress, validatorAddress, amount, }) {
        this.delegatorAddress = delegatorAddress;
        this.validatorAddress = validatorAddress;
        this.amount = amount;
    }
    toProto() {
        return __awaiter(this, void 0, void 0, function* () {
            const msgContent = {
                delegatorAddress: this.delegatorAddress,
                validatorAddress: this.validatorAddress,
                amount: this.amount,
            };
            return {
                typeUrl: `/cosmos.staking.v1beta1.MsgUndelegate`,
                value: msgContent,
                encode: () => __awaiter(this, void 0, void 0, function* () {
                    return (yield Promise.resolve().then(() => __importStar(require("../protobuf_stuff/cosmos/staking/v1beta1/tx")))).MsgUndelegate.encode(msgContent).finish();
                }),
            };
        });
    }
    toAmino() {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                type: "cosmos-sdk/MsgUndelegate",
                value: {
                    delegator_address: this.delegatorAddress,
                    validator_address: this.validatorAddress,
                    amount: this.amount,
                },
            };
        });
    }
}
exports.MsgUndelegate = MsgUndelegate;
